From bfdaccaa9a3ba1b5a29c42fb1780b3af418a0886 Mon Sep 17 00:00:00 2001
From: Duy Truong <duytruong@google.com>
Date: Sun, 14 Jun 2026 07:44:28 +0000
Subject: [PATCH 2/2] Fix out-of-bounds write in stpropnci_process

In the ST NFC HAL, the `INfc::write()` implementation lacked bounds checking for incoming payloads. This could allow a compromised client to send a payload exceeding `MAX_NCI_MESSAGE_LEN`, triggering an unbounded `memcpy` into a fixed-size buffer. This results in an out-of-bounds write over global state variables, potentially leading to arbitrary code execution.

This commit resolves the vulnerability by adding a bounds check in `stpropnci_process()` to drop oversized payloads early and log an error if `payloadlen` exceeds `MAX_NCI_MESSAGE_LEN`.

Test: Built libstpropnci-vendor for cf_x86_64_phone-trunk_staging-userdebug and executed alint.

Bug: 522376740
Flag: EXEMPT BUGFIX
This was autogenerated by MendIt (go/androidmendit)

(cherry picked from commit 867b2f4b7d87d95ffdfc4e070f42a117a6b87fdd)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:003c23feb1d99e338ce4457a17a7c2b83d33027f
Merged-In: I492e6cbca61f895f75d25bee67859047d9734523
Change-Id: I492e6cbca61f895f75d25bee67859047d9734523
---
 libstpropnci/src/stpropnci.cc | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/libstpropnci/src/stpropnci.cc b/libstpropnci/src/stpropnci.cc
index abcce4e..ec1c498 100755
--- a/libstpropnci/src/stpropnci.cc
+++ b/libstpropnci/src/stpropnci.cc
@@ -230,6 +230,12 @@ bool stpropnci_process(bool dir_from_upper, const uint8_t* payload,
     (void)pthread_mutex_unlock(&reentry_lock);
     return ret;
   }
+  if (payloadlen > MAX_NCI_MESSAGE_LEN) {
+    LOG_E("Dropping oversized NCI message: %d > %d", payloadlen,
+          MAX_NCI_MESSAGE_LEN);
+    (void)pthread_mutex_unlock(&reentry_lock);
+    return ret;
+  }
   if (!stpropnci_state.passthrough_mode) {
     if (payloadlen > 3) {
       LOG_D("Processing (hdr:%02hhx%02hhx%02hhx%02hhx)", payload[0], payload[1],
-- 
2.55.0

